server/authz/policy.rego

```diff
@@ -333,20 +351,38 @@ allow {
 
 # (Team) observers can run only if observers_can_run
 allow {
-	object.type == "query"
-	object.observer_can_run == true
-	subject.global_role == observer
-	action = run
+  object.type == "targeted_query"
+  object.observer_can_run == true
+  subject.global_role == observer
+  action = run
 }
+
 # Team observer running a observers_can_run query must have the targets
-# filtered to only teams that they observe. That check is not validated by this rego
-# file, it is a filter that is applied at the datastore level (in HostIDsInTargets).
-allow {
-	object.type == "query"
-	object.observer_can_run == true
-	# If role is observer on any team
-	team_role(subject, subject.teams[_].id) == observer
-	action == run
+# filtered to only teams that they observe.
+allow {
+  object.type == "targeted_query"
+  object.observer_can_run == true
+  is_null(subject.global_role)
+  action == run
+
+  not is_null(object.host_targets.teams)
+  ok_teams := { tmid | tmid := object.host_targets.teams[_]; team_role(subject, tmid) == [admin,maintainer,observer][_] }
+  count(ok_teams) == count(object.host_targets.teams)
+}
+
+# Team observer running a observers_can_run query and there are no
+# target teams.
+allow {
+  object.type == "targeted_query"
+  object.observer_can_run == true
+  is_null(subject.global_role)
+  action == run
+
+  # If role is admin, maintainer or observer on any team
+  team_role(subject, subject.teams[_].id) == [admin,maintainer,observer][_]
+
+  # and there are no team targets
+  is_null(object.host_targets.teams)
 }
```

Comment:
Note that it's not just the true/false of `observer_can_run` that changes, the list of team roles allowed is different. While I'm sure there's a way to parameterize this in rego, I think there's value in keeping it as straightforward as possible because that language quickly gets confusing and we'll likely touch it rarely enough to never learn it in a way that sticks.